### PR TITLE
Remote CKAN config

### PR DESCRIPTION
**Pull request by amercader**
See https://github.com/ckan/ideas-and-roadmap/pull/145 for context (now a bit outdated, will update it next week)

Tl;DR of this PR:

```
http POST localhost:5000/api/action/config_option_update Authorization:XXX ckan.homepage_style=2 ckan.site_title="My site title" ckan.site_description="Nice portal" ckanext.example_iconfigurer.test_conf="test_value" ckan.datasets_per_page=5

{
    "help": "http://localhost:5000/api/3/action/help_show?name=config_option_update", 
    "result": {
        "ckan.datasets_per_page": 5, 
        "ckan.homepage_style": 2, 
        "ckan.site_description": "Nice portal", 
        "ckan.site_title": "My site title", 
        "ckanext.example_iconfigurer.test_conf": "test_value"
    }, 
    "success": true
}

```

And also:

zraunid

A bit more detail:

* Refactored the code that handles storing settings in the db (and on the `app_globals` object) to be a bit more sane, to use actions and check which config options can be updated using a schema (which is also used to validate them)
* Extensions can modify this schema to add (or remove) config options to it, making them available to update ([example](https://github.com/ckan/ckan/pull/2429/files#diff-8d8bfa5ed17aef6869bd49c39aad1e3e))
* The `config_option_update` action can be used to edit any of the allowed options in bulk passing config key and values, as in the example above.
* `config_option_list`will return the available options, and `config_option_show` the current value for one of them
* Extensions can also extend the admin config form to allow editing the configuration options from the frontend, as in the screenshot above.

### CODE REVIEW

File: ckan/logic/action/update.py [hunk 1]
```diff
@@ -1308,3 +1311,87 @@ def bulk_update_delete(context, data_dict):
...
     _check_access('bulk_update_delete', context, data_dict)
     _bulk_update_dataset(context, data_dict, {'state': 'deleted'})
+
+
+def config_option_update(context, data_dict):
+    '''
+
+    .. versionadded:: 2.4
+
+    Allows to modify some CKAN configuration options
```
Comment (davidread) [inline ckan/logic/action/update.py]: I think this is clearer as: "Modify the values of one or more CKAN configuration options"

File: ckan/lib/app_globals.py [hunk 2]
```diff
@@ -20,19 +22,17 @@
...
-    'ckan.site_description',
-    'ckan.site_about',
-    'ckan.site_intro_text',
-    'ckan.site_custom_css',
-    'ckan.homepage_style',
-]
-
-config_details = {
+
+# This mapping is only used to define the configuratin options (from the
```
Comment (davidread) [inline ckan/lib/app_globals.py]: typo "configuratin"

File: doc/extensions/remote-config-update.rst [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+`configuration file <CKAN configuration file>`_ and restart the server.
+
+.. warning::
+
+    Only configuration options which are not critical, sensitive or could cause the
+    CKAN instance to break should be made available.
+
+.. note::
+
+    Only sysadmin users are allowed to modify runtime configuratin options.
```
Comment (davidread) [inline doc/extensions/remote-config-update.rst]: typo 'configuratin'

File: ckan/logic/action/update.py [hunk 4]
```diff
@@ -1308,3 +1311,87 @@ def bulk_update_delete(context, data_dict):
...
+      options explicitly defined in the ``app_globals`` module.
+
+    The following lists a ``key`` parameter, but this should be replaced by
+    whichever config options want to be updated, eg::
+
+        get_action('config_option_update)({}, {
+            'ckan.site_title': 'My Open Data site',
+            'ckan.homepage_layout': 2,
+        })
+
```
Comment (davidread) [inline ckan/logic/action/update.py]: How about adding a warning like this: You should only add config options that you know in the code are ok to be updated during runtime, such as ones you've added in your own extension, or have reviewed the use of in core ckan.

**Comment by davidread** [comment]
Language-wise, I'm concerned that the config options that are stored in system_info are called a few different things:
* config_option (e.g. in the action function names)
* editable
* can be edited/modified/updated at runtime

How about we just call them "runtime-editable config options" everywhere?

Secondly, can you say "during runtime" rather than "at runtime", as the latter seems to me it might mean it is set at the moment when you start the server.

File: doc/maintaining/configuration.rst [hunk 5]
```diff
@@ -1,15 +1,65 @@
...
+on the server operating system.
+
+These are generally low-level critical settings needed when setting up the application, like the database
+connection, the Solr server URL, etc. Sometimes it can be useful to define them as environment variables to
+automate and orchestrate deployments without having to first modify the `configuration file <CKAN configuration file>`_.
+
+These options are only read at startup time to update the ``config`` object used by CKAN,
+but they won't we accessed any more during the lifetime of the application.
+
+CKAN environment variables match the options in the configuration file, but they are always uppercase
```
Comment (davidread) [inline doc/maintaining/configuration.rst]: CKAN environment variables NAMES match the options

File: doc/maintaining/configuration.rst [hunk 6]
```diff
@@ -1,15 +1,65 @@
...
+
+Updating configuration options at runtime
+*****************************************
+
+CKAN configuration options are generally defined before starting the web application (either in the
+`configuration file <CKAN configuration file>`_ or via `Environment variables`_).
+
+A limited number of configuration options can also be set during runtime. This can be done on the
+:ref:`administration interface <admin page>` or using the :py:func:`~ckan.logic.action.update.config_option_update`
+API action. Only :doc:`sysadmins </sysadmin-guide>` can edit these configuration options. Changes made to these configuation options will be stored on the database and persisted when the server is restarted.
```
Comment (davidread) [inline doc/maintaining/configuration.rst]: spelling typo: configuation

File: ckan/controllers/admin.py [hunk 7]
```diff
@@ -51,6 +56,15 @@ def _get_config_form_items(self):
         return items
 
     def reset_config(self):
+        '''FIXME: This method is probably not doing what people would expect.
+           It will reset the configuration to values cached when CKAN started.
+           If these were coming from the database during startup, that's the
+           ones that will get applied on reset, not the ones on the ini file.
```
Comment (davidread) [inline ckan/controllers/admin.py]: I think "in the ini file" is more natural than "on". There are a few of these.

**Comment by amercader** [comment]
@davidread I think I've covered all cases to use "runtime-editable config options" consistently, as well as the rest of your points. Let me know if there is anything else.

File: ckan/model/system_info.py [hunk 8]
```diff
@@ -1,29 +1,45 @@
 from sqlalchemy import types, Column, Table
 
+import vdm.sqlalchemy
 import meta
 import core
 import domain_object
 
 __all__ = ['system_info_revision_table', 'system_info_table', 'SystemInfo',
```
Comment (davidread) [inline ckan/model/system_info.py]: It would be good to state clearly here that "SystemInfo" = "run-time editable config option"

File: doc/extensions/remote-config-update.rst [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+.. warning::
+
+    Only configuration options which are not critical, sensitive or could cause the
+    CKAN instance to break should be made runtime-editable. You should only add config options that you are comfortable
+    they can be edited during runtime, such as ones you've added in your
+    own extension, or have reviewed the use of in core CKAN.
+
+.. note::
+
+    Only sysadmin users are allowed to modify runtimei-editable configuratin options.
```
Comment (davidread) [inline doc/extensions/remote-config-update.rst]: typos 'runtimei' 'configuratin'

**Comment by davidread** [comment]
I really appreciate this little improvements. Just made a couple more suggestions and I think that's it.

**Comment by amercader** [comment]
@davidread done. TBH I'm not sure why this table was called `system_info` when it was first implemented, perhaps it was meant to be a more generic key/value store for internal CKAN functions. In any case I agree that is best to make clear why we are actually using it now.
